src/java/org/apache/cassandra/index/sai/disk/v1/MergePostingList.java

```diff
@@ -8,334 +7,125 @@
         this.size = size;
     }
 
     public static PostingList merge(List<PostingList> postings, Closeable onClose)
     {
         checkArgument(!postings.isEmpty());
-        return postings.size() > 1 ?
-               new MergePostingList(postings, onClose) : postings.get(0);
+        return postings.size() > 1 ? new MergePostingList(postings, onClose) : postings.get(0);
     }
 
     public static PostingList merge(List<PostingList> postings)
     {
         return merge(postings, () -> FileUtils.close(postings));
     }
 
-    @Override
-    public void close() throws IOException
-    {
-        onClose.close();
-    }
-
-    @Override
-    public int size()
-    {
-        return size;
-    }
-
     @SuppressWarnings("resource")
     @Override
-    public final int nextPosting() throws IOException
+    public int nextPosting() throws IOException
     {
-        for (int i = dirtyIdx; i > 0; i--)
+        while (!postingLists.isEmpty())
         {
-            final PeekingPostingList list = dirty[i - 1];
-            int peek = list.peek();
-            // we're removing duplicates
-            if (peek <= lastRowId)
+            PeekablePostingList head = postingLists.poll();
+            int next = head.nextPosting();
+
+            if (next == END_OF_STREAM)
             {
-                list.consume();
-                i++;
-                continue;
+                // skip current posting list
             }
-            else if (peek != PostingList.END_OF_STREAM)
+            else if (next > lastRowId)
             {
-                // only add back to the heap if there's anything left in the list
-                sorted.add(list);
+                lastRowId = next;
+                postingLists.add(head);
+                return next;
+            }
+            else if (next == lastRowId)
+            {
+                postingLists.add(head);
             }
-            dirtyIdx--;
-        }
-
-        final PeekingPostingList min = sorted.poll();
-        if (min == null)
-        {
-            return PostingList.END_OF_STREAM;
-        }
-
-        int leadingIdx = 0;
-        leading[leadingIdx++] = min;
-
-        // Record the other list with the same current ID...
-        while (!sorted.isEmpty() && sorted.peek().currentRowID == min.currentRowID)
-        {
-            leading[leadingIdx++] = sorted.poll();
-        }
-
-        for (int idx = 0; idx < leadingIdx; ++idx)
-        {
-            PeekingPostingList list = leading[idx];
-            // just consumes buffered top value
-            list.consume();
-            dirty[dirtyIdx++] = list;
         }
 
-        return lastRowId = min.currentRowID;
+        return PostingList.END_OF_STREAM;
     }
 
+    @SuppressWarnings("resource")
     @Override
     public int advance(int targetRowID) throws IOException
     {
-        int nextMin = PostingList.END_OF_STREAM;
-        int leadingIdx = 0;
-        for (PeekingPostingList list : peekingPostingLists)
```

Comment:
nit: Not necessarily needed now with the current usage of `temp`, but I'm wondering if adding a `temp.clear();` statement here before potentially adding to it will safeguard us in the future?

Reply:
+1